### DISCUSSION TITLE
Somehow get notified on "--watch" or "--hot" events

### DISCUSSION THREAD

**Discussion by i9or** [oven-sh/bun#10826] [Q&A]
Hi, not sure what I'm trying to solve is actually possible. I am building a simple static HTML web server, using Bun.serve and Hono framework. I would like to add simple reload on source code change (or specifically on templates folder changes). Currently I was able to make a hacky solution, where any open page just reloads on WebSocket connection lost event in 500ms. It's kind of enough for now, but once server reload is above 500ms, it will start hurting DX:

```ts
// server.ts
const server = Bun.serve({
  fetch: app.fetch,
  port: 3000,
});

const wss = new WebSocketServer({ port: 3001 });

console.info(`🌭 Listening on ${server.url}`);
```

```ts
// Layout.ts
export const Layout = ({children}) => {
  return html`<!doctype html><html><body>${children}<script>
    const ws = new WebSocket("ws://localhost:1338");

    ws.onclose = function () {
      console.info("[DEV] Connection lost, reloading in 500ms...");
      setTimeout(() => {
        window.location.reload();
      }, 500);
    };
  </script>
</body></html>`;
};
```

Are there any other approaches? Maybe there's a hidden callback or event I can subscribe to with other process?

**Comment by truongan07** [comment]
Why does it reload above 500ms ? you think timeout was changed by someone from frontend ?

**Reply by i9or** [reply]
It is not at the moment, but I feel like adding more things to the server code might increase the full reload time above 500ms. On the opposite, I would like reload in the browser happening right after reload on the server is finished, right now it's around tens of milliseconds. Either way I need some notification from the Bun that it finished reloading the code to trigger browser reload without any artificial delay.

**Reply by truongan07** [reply]
You use --hot flag or -- watch flag ?

**Reply by i9or** [reply]
I currently use `--watch` flag, as it does the full reload. Tried to use `--hot` flag, but haven't found any way to trigger reload in the browser, as there's no way to determine hot reload have finished. Or at least I am not aware of any.

**Reply by truongan07** [reply]
You should use --hot flag to reflect the updated code without restarting the process. Example : 
```
// server.ts
const server = Bun.serve({
  fetch: app.fetch,
  port: 3000,
});

const wss = new WebSocketServer({ port: 3001 });

wss.onopen = function () {
    if(!(globalThis.reloading instanceof Object)) {
        globalThis.reloading = { isHot: false, count: 0 };
    }
};

console.info(`🌭 Listening on ${server.url}`);
```
 ```
// Layout.ts
let timeout = 500;

if(globalThis.reloading instanceof Object) if(globalThis.reloading?.isHot) {
    // you can use counter if necessary
    ++globalThis.reloading.count;

    // do something with --hot flag. Example reloading in 0ms
    timeout = 0;
}else globalThis.reloading.isHot = true;

export const Layout = ({children}) => {
  return html`<!doctype html><html><body>${children}<script>
    const ws = new WebSocket("ws://localhost:1338");

    ws.onclose = function () {
        console.info(`[DEV] Connection lost, reloading in ${timeout}ms...`);
        setTimeout(() => {
            window.location.reload();
        }, timeout);
    };
  </script>
</body></html>`;
};
```

**Reply by truongan07** [reply]
> but haven't found any way to trigger reload in the browser

--watch & --hot flag for reloading on server without reload in the browser at client. You need to use other technicality to do this such as set value of cookie is true when reload on server by --hot flag and check this cookie by interval method at client. if is true, will call reload and set value of cookie is false

**Reply by i9or** [reply]
Thanks for trying to help, unfortunatelly none of the above is relevant to my question.

**Comment by vitch** [comment]
I have a similar question...

In the frontend of my project I can use `server.watcher.on('change'` from `vite` to be notified whenever vite's code watcher has decided there has been a change and the code needs to be reloaded.

For the backend portion (run with `bun --hot`) I want something similar but I can't find anything analogous. I basically need to know the code changed in order to reset some caches...

**Reply by RiskyMH** [reply]
For frontend (ie `bun index.html`), bun supports the import.meta.hot api from vite https://bun.sh/docs/bundler/hmr#import-meta-hot-on-and-off

Would be cool if it was got more though (I don't think it works for server --hot)

**Reply by ajitid** [reply]
@RiskyMH Can we raise a feature request so that `import.meta.hot` API could be added for server --hot too? 
I too am using global counter for now to observe when hot reloads are made.
